File: C1.UWP/CS/BasicLibrarySamples/Utils/DelegateCommand.cs
using System;
using System.Windows.Input;

namespace BasicLibrarySamples
{
    public class DelegateCommand<T> : ICommand
    {
        private static readonly Action<T> EmptyExecute = (T) => { };
        private static readonly Func<T, bool> EmptyCanExecute = (T) => true;


        private Action<T> _execute;
        private Func<T, bool> _canExecute;

        public DelegateCommand(Action<T> executeMethod)
            : this(executeMethod, null)
        {
        }

        public DelegateCommand(Action<T> execute, Func<T, bool> canExecute)
        {
            _execute = execute;
            _canExecute = canExecute;
        }

        public void Execute(T parameter)
        {
           _execute(parameter);
        }


        public bool CanExecute(T parameter)
        {
            return _canExecute(parameter);
        }


        bool ICommand.CanExecute(object parameter)
        {
            return this.CanExecute((T)parameter);
        }


        public event EventHandler CanExecuteChanged;
        public void RaiseCanExecuteChanged()
        {
            var h = this.CanExecuteChanged;
            if (h != null)
            {
                h(this, EventArgs.Empty);
            }
        }


        void ICommand.Execute(object parameter)
        {
            this.Execute((T)parameter);
        }
    }

}

File: C1.UWP/VB/BasicLibrarySamples/Utils/DelegateCommand.vb
Imports System.Windows.Input
Imports System

Public Class DelegateCommand(Of T)
    Implements ICommand
    Private Shared ReadOnly EmptyExecute As Action(Of T) = Nothing
    Private Shared ReadOnly EmptyCanExecute As Func(Of T, Boolean) = Function(T) True


    Private _execute As Action(Of T)
    Private _canExecute As Func(Of T, Boolean)

    Public Sub New(executeMethod As Action(Of T))
        Me.New(executeMethod, Nothing)
    End Sub

    Public Sub New(execute As Action(Of T), canExecute As Func(Of T, Boolean))
        _execute = execute
        _canExecute = canExecute
    End Sub

    Public Sub Execute(parameter As T)
        _execute(parameter)
    End Sub


    Public Function CanExecute(parameter As T) As Boolean
        Return _canExecute(parameter)
    End Function


    Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return Me.CanExecute(CType(parameter, T))
    End Function


    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

    Public Sub RaiseCanExecuteChanged()
        RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
    End Sub


    Sub Execute(parameter As Object) Implements ICommand.Execute
        Me.Execute(CType(parameter, T))
    End Sub
End Class
